service/lib/agama/dbus/base_tree.rb

```diff
@@ -0,0 +1,153 @@
+# frozen_string_literal: true
+
+# Copyright (c) [2023] SUSE LLC
+#
+# All Rights Reserved.
+#
+# This program is free software; you can redistribute it and/or modify it
+# under the terms of version 2 of the GNU General Public License as published
+# by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "dbus"
+
+module Agama
+  module DBus
+    # Base class for a D-Bus tree exporting a collection of D-Bus objects
+    #
+    # Derived classes must define these methods:
+    #   * {BaseTree#create_dbus_object}
+    #   * {BaseTree#update_dbus_object}
+    #   * {BaseTree#dbus_object?}
+    class BaseTree
+      # Constructor
+      #
+      # @param service [::DBus::ObjectServer]
+      # @param root_path [::DBus::ObjectPath] Root path of the tree
+      # @param logger [Logger, nil]
+      def initialize(service, root_path, logger: nil)
+        @service = service
+        @root_path = root_path
+        @logger = logger
+      end
+
+      # Adds, updates or deletes D-Bus objects according to the given list of objects
+      #
+      # As result there is an exported D-Bus object for each object in the list.
+      #
+      # @param objects [Array]
+      def objects=(objects)
+        add_objects(objects)
+        update_objects(objects)
+        delete_objects(objects)
+      end
+
+    private
+
+      # @return [::DBus::ObjectServer]
+      attr_reader :service
+
+      # @return [::DBus::ObjectPath]
+      attr_reader :root_path
+
+      # @return [Logger]
+      attr_reader :logger
+
+      # Returns the D-Bus object for the given object
+      #
+      # @param object [Object]
+      # @return [DBus::BaseObject, nil]
+      def find_dbus_object(object)
+        dbus_objects.find { |o| dbus_object?(o, object) }
+      end
+
+      # Creates a D-Bus object for the given object
+      #
+      # @raise Must be defined by derived classes
+      #
+      # @param _object [Object]
+      # @return [DBus::BaseObject]
+      def create_dbus_object(_object)
+        raise "Undefined by derived class"
```

Comment:
```suggestion
        raise "Abstract method, derived classes must implement this"
```

see also https://rubydoc.info/gems/yard/file/docs/Tags.md#abstract